porcelain/storagedeals.go

```diff
@@ -1,25 +1,87 @@
 package porcelain
 
 import (
+	"context"
+
 	"github.com/ipfs/go-cid"
 
+	"github.com/filecoin-project/go-filecoin/plumbing/strgdls"
 	"github.com/filecoin-project/go-filecoin/protocol/storage/storagedeal"
 )
 
-type strgdlsPlumbing interface {
-	DealsLs() ([]*storagedeal.Deal, error)
+type dealGetPlumbing interface {
+	DealsLs(context.Context) (<-chan *strgdls.StorageDealLsResult, error)
 }
 
 // DealGet returns a single deal matching a given cid or an error
-func DealGet(plumbing strgdlsPlumbing, dealCid cid.Cid) *storagedeal.Deal {
-	deals, err := plumbing.DealsLs()
+func DealGet(ctx context.Context, plumbing dealGetPlumbing, dealCid cid.Cid) *storagedeal.Deal {
+	dealCh, err := plumbing.DealsLs(ctx)
 	if err != nil {
 		return nil
 	}
-	for _, storageDeal := range deals {
-		if storageDeal.Response.ProposalCid == dealCid {
-			return storageDeal
+	for deal := range dealCh {
+		if deal.Err != nil {
+			return nil
+		}
+		if deal.Deal.Response.ProposalCid == dealCid {
+			return &deal.Deal
 		}
 	}
 	return nil
 }
+
+type dealClientLsPlumbing interface {
+	ConfigGet(string) (interface{}, error)
+	DealsLs(context.Context) (<-chan *strgdls.StorageDealLsResult, error)
+}
+
+// DealClientLs returns a channel with all deals placed as a client
+func DealClientLs(ctx context.Context, plumbing dealClientLsPlumbing) (<-chan *strgdls.StorageDealLsResult, error) {
+	results := make(chan *strgdls.StorageDealLsResult)
+
+	minerAddress, _ := plumbing.ConfigGet("mining.minerAddress")
+
+	dealCh, err := plumbing.DealsLs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		for deal := range dealCh {
+			if deal.Err != nil || deal.Deal.Miner != minerAddress {
+				results <- deal
+			}
+		}
+		close(results)
+	}()
+
+	return results, nil
+}
+
+type dealMinerLsPlumbing interface {
+	ConfigGet(string) (interface{}, error)
+	DealsLs(context.Context) (<-chan *strgdls.StorageDealLsResult, error)
+}
+
+// DealMinerLs returns a channel with all deals received as a miner
+func DealMinerLs(ctx context.Context, plumbing dealMinerLsPlumbing) (<-chan *strgdls.StorageDealLsResult, error) {
+	results := make(chan *strgdls.StorageDealLsResult)
+
+	minerAddress, _ := plumbing.ConfigGet("mining.minerAddress")
+
+	dealCh, err := plumbing.DealsLs(ctx)
+	if err != nil {
+		return results, err
```

Comment:
Why does this return results, err and the other return `nil, err` in the same place? 

This code is identical except for names and where the deal miner is or is not equal to the minerAddress.  Is there no way to DRY it up?  I'm sure you tried, just want to know what stopped it.

Reply:
Also I had to extract the code into files and do a diff to understand what was different; this is a code smell to me, because it will tempt somebody to unify it, and while they'll surely fail, their time will have been blown understanding it.  I suggest trying to extract some of this into smaller function calls for readability, and/or comment.

Reply:
The `return results, err` was a mistake. I've fixed it.

I also extracted some of the common functionality to DRY up the code.